jax/experimental/mosaic/gpu/examples/fdgg.py

```diff
@@ -0,0 +1,505 @@
+                                         cx(n_tile_count))
+      work_id_group_end = arith.addi(work_id_group_start, group_work_item_count)
+
+      @subgroup_for(work_id, work_id_group_end, worker_count)
+      def subgroup_for_body(work_id):
+        tile_n_start = arith.muli(arith.remui(work_id, cx(n_tile_count)),
+                                  cx(tile_n))
+        work_id_within_group = arith.subi(work_id, work_id_group_start)
+        subgroup = arith.divui(work_id_within_group, cx(n_tile_count))
+        # TODO(andportnoy) try to pick better names for all these variables
+        #                  and streamline layout calculations in general
+        tile_m_start_within_group = arith.muli(subgroup, cx(tile_m))
+        tile_m_start = arith.addi(group_offset, tile_m_start_within_group)
+        tile_m_end_within_group = arith.minui(arith.addi(tile_m_start_within_group, cx(tile_m)),
+                                              group_size)
+        tile_m_end = arith.addi(group_offset, tile_m_end_within_group)
+        tile_m_count = arith.subi(tile_m_end, tile_m_start)
+
+        # output n_start
+        o_n_start = tile_n_start
+
+        # weight n_start
+        w_n_start = arith.addi(arith.muli(expert_id, cx(expert_n)),
+                               tile_n_start)
+
+        local_work_id = arith.divui(work_id, worker_count)
+        tmem_slot = arith.remui(local_work_id, cx(tmem_slot_count))
+
+        tmem_address = arith.addi(acc.address,
+                                  arith.index_cast(i32,
+                                                   arith.muli(tmem_slot, cx(tile_n))))
+        acc_slot = tcgen05.TMEMRef(
+            address=tmem_address,
+            shape=(tile_m, tile_n),
+            layout=acc.layout,
+            dtype=acc.dtype,
+        )
+        get_persistent_ki = lambda ki: arith.addi(ki, arith.muli(local_work_id, cx(k_loop_iter)))
+        with only(tma_warp), single_warp_thread():
+          @mgpu.fori(cx(k_loop_iter), None)
+          def _tma_body(ki, _):
+            persistent_ki = get_persistent_ki(ki)
+            slot = arith.remui(persistent_ki, c(max_concurrent_steps, index))
+            with ctx.named_region("TMA wait for MMA"):
+              with mgpu.when(arith.cmpi(arith.CmpIPredicate.uge, persistent_ki, c(max_concurrent_steps, index))):
+                ab_empty_barriers[slot].wait()
+            full_barrier = ab_full_barriers[slot]
+            full_barrier.arrive_expect_tx(
+                bytecount((tile_m, tile_k), in_dtype) + bytecount((tile_n, tile_k), in_dtype)
+            )
+            k_start = arith.muli(ki, c(tile_k, index))
+            common_args = dict(
+                swizzle=swizzle,
+                barrier=full_barrier,
+                arrive=False,
+                uniform=False,
+            )
+
+            a_smem_slot = mgpu.memref_slice(a_smem, slot)
+            a_smem_slot_2d_shape, _ = mma_utils.tiled_memref_shape(a_smem_slot)
```

Comment:
Yeah you should not use mma_utils in this kernel. It's a private file. What's happening here?